*: add best practices for highly concurrent write scenario (#1630)
pingcap/docs: dev/reference/best-practices/high-concurrency.md

During the continuous write process, after finding that hotspot is caused on Node 1, PD evenly distributes the concentrated Leaders to other nodes. If the number of TiKV nodes is more than the number of Region replicas, TiKV will try to migrate these Regions to idle nodes. These two operations during the write process are also reflected in the PD's monitoring metrics:

![QPS5](/media/best-practices/QPS5.png)

After a period of continuous writes, PD automatically schedules the entire TiKV cluster to a state where pressure is evenly distributed. By that time, the capacity of the whole cluster can be fully used.

In most cases, the above process of causing a hotspot is normal, which is the Region warm-up phase of database. However, you need to avoid this phase in highly-concurrent write-intensive scenarios.

> ## Hotspot solutions

Reviewer: It seems that we only provide one solution.
```suggestion
## Hotspot solution
```

To achieve the ideal performance expected in the theoretical scenario, one method is to skip the warm-up phase by directly splitting a Region into the desired number of Regions and scheduling these Regions in advance to other nodes in the cluster.

In v3.0.x and from v2.1.13 on, TiDB supports a new feature called [Split Region](/dev/reference/sql/statements/split-region.md). This new feature provides the following new syntaxes:

{{< copyable "sql" >}}

```sql
SPLIT TABLE table_name [INDEX index_name] BETWEEN (lower_value) AND (upper_value) REGIONS region_num
```

{{< copyable "sql" >}}

```sql
SPLIT TABLE table_name [INDEX index_name] BY (value_list) [, (value_list)]
```

However, TiDB does not automatically perform this pre-split operation. The reasons are related to the data distribution in TiDB as one of its feature:

![Table Region Range](/media/best-practices/table-Region-range.png)

> As can be seen from the image above, the key of the row data is so encoded that `rowID` is the only variable. In TiDB, `rowID` is an `Int64` integer. However, the user might not need to evenly split the `Int64` integer range to the desired number of ranges and then to distribute these ranges to different nodes, because Region split must also be based on the actual situation.

Reviewer: @winkyao This sentence (`Table 行数据 key 的编码中`) is confusing to me. PTAL

```
[root@172.16.4.4 scripts]# python table-regions.py --host 172.16.4.3 --port 31453 test test_hotspot
[RECORD - test.test_hotspot] - Leaders Distribution:
  total leader count: 127
  store: 1, num_leaders: 21, percentage: 16.54%
  store: 4, num_leaders: 20, percentage: 15.75%
  store: 6, num_leaders: 21, percentage: 16.54%
  store: 46, num_leaders: 21, percentage: 16.54%
  store: 82, num_leaders: 23, percentage: 18.11%
  store: 62, num_leaders: 21, percentage: 16.54%
```

Then operate the write load again:

![QPS6](/media/best-practices/QPS6.png)

![QPS7](/media/best-practices/QPS7.png)

![QPS8](/media/best-practices/QPS8.png)

You can see that the hotspot has been eliminated now.

In this case, the table is simple. In other cases, you might also need to consider the hotspot problem of index. For more details on how to pre-split the index Region, refer to [Split Region](/dev/reference/sql/statements/split-region.md).

## More complex hotspot problem

> If a table does not have a primary key, or the primary key is not the `Int` type and you do not want to generate a randomly distributed primary key ID, TiDB provides an implicit `_tidb_rowid` column as the row ID. If you do not use the `SHARD_ROW_ID_BITS` parameter, the values of the `_tidb_rowid` column might be monotonically increasing, which also causes hotspot. Refer to [`SHARD_ROW_ID_BITS` description](/dev/reference/configuration/tidb-server/tidb-specific-variables.md#shard_row_id_bits) for more detail.

Reviewer: ```suggestion
If a table does not have a primary key, or the primary key is not the `Int` type and you do not want to generate a randomly distributed primary key ID, TiDB provides an implicit `_tidb_rowid` column as the row ID. Generally, when you do not use the `SHARD_ROW_ID_BITS` parameter, the values of the `_tidb_rowid` column are also monotonically increasing, which might causes hotspots too. Refer to [`SHARD_ROW_ID_BITS` description](/dev/reference/configuration/tidb-server/tidb-specific-variables.md#shard_row_id_bits) for more details.
```

> To avoid the hotspot problem in this situation, you can use `SHARD_ROW_ID_BITS` and `PRE_SPLIT_REGIONS` when creating a table. For more details about `PRE_SPLIT_REGIONS`, refer to [here](/dev/reference/sql/statements/split-region.md#pre_split_regions).

Reviewer: Avoid `here`. Use the specific description instead.

```suggestion
To avoid the hotspot problem in this situation, use `SHARD_ROW_ID_BITS` and `PRE_SPLIT_REGIONS` when creating a table. For more details about `PRE_SPLIT_REGIONS`, refer to [Pre-split regions](/dev/reference/sql/statements/split-region.md#pre_split_regions).
```

`SHARD_ROW_ID_BITS` is used to randomly scatter the row ID generated in the `_tidb_rowid` column. `pre_split_regions` is used to pre-split the Region after a table is created.

> **Note:**
>
> The value of `pre_split_regions` must be smaller or equal to that of `shard_row_id_bits`.

Example:

{{< copyable "sql" >}}

```sql
create table t (a int, b int) shard_row_id_bits = 4 pre_split_regions=·3;
```

- `SHARD_ROW_ID_BITS = 4` means that the values of `tidb_rowid` will be randomly distributed into 16 (16=2^4) ranges.
- `pre_split_regions=3` means that the table will be pre-split into 8 (2^3) Regions after it is created.

> When data starts to be written into table t, the data is written into the pre-split 8 Regions, which avoids the hotspot problem that might be caused if there is only one Region after the table creation.

Reviewer: ```suggestion
When data starts to be written into table `t`, the data is written into the pre-split 8 Regions, which avoids the hotspot problem that might be caused if only one Region exists after table creation.
```

## Parameter configuration

> In v2.1, the [latch mechanism](/dev/reference/configuration/tidb-server/configuration-file.md#txn-local-latches) is introduced in TiDB to identify transaction conflicts in advance in scenarios where write conflicts frequently appear. The aim is to reduce the retry of transaction commits in TiDB and TiKV caused by write conflicts. Generally, batch tasks use the data already stored in TiDB, so the write conflicts of transaction do not exist. In this situation, you can disable the latch in TiDB to reduce memory allocation for small objects:

Reviewer: ```suggestion
In v2.1, the [latch mechanism](/dev/reference/configuration/tidb-server/configuration-file.md#txn-local-latches) is introduced in TiDB to identify transaction conflicts in advance in scenarios where write conflicts frequently appear. The aim is to reduce the retry of transaction commits in TiDB and TiKV caused by write conflicts. Generally, batch tasks use the data already stored in TiDB, so write conflicts of transactions do not occur. In this situation, you can disable the latch in TiDB to reduce memory allocation for small objects:
```

Reviewer: `to reduce memory allocation for small objects`: I'm not sure about this description and the Chinese version is also confusing to me `以减少细小内存对象的分配`. @winkyao PTAL
